Stripe/STPPaymentContext.m

```diff
@@ -554,6 +579,103 @@ - (void)requestSourceFlowNonePayment:(STPSource *)source {
+                                                                              returnURL:self.configuration.returnURLBlock().absoluteString
+                                                                                   card:cardSource.stripeID];
+
+    STPSourceCompletionBlock threeDSSourceCompletion = ^(STPSource * _Nullable threeDSSource, NSError * _Nullable error) {
+        if (error
+            || threeDSSource == nil
+            || threeDSSource.flow != STPSourceFlowRedirect) {
+            if (!threeDSecureRequired
+                && error.domain == StripeDomain
+                && error.code == STPCardError
+                && [error.userInfo[STPCardErrorCodeKey] isEqualToString:STPPaymentMethodNotAvailable]) {
+                // This is a card that doesn't actually support 3ds and the card source didn't say 3ds was required
+                // So we can just return the original card source to be charged
+                self.state = STPPaymentContextStateNone;
+                [self requestSynchronousSourcePayment:cardSource];
+
+            }
+            else {
+                // Finish with an error
+                [self didFinishWithStatus:STPPaymentStatusError
+                                    error:error ?: [NSError stp_genericConnectionError]];
+            }
+        }
+        else {
+            // We successfully have a 3ds source
+            switch (threeDSSource.status) {
+                case STPSourceStatusPending:
+                    // Do 3DS redirect
+                    self.state = STPPaymentContextStateNone;
+                    [self requestSourceFlowRedirectPayment:threeDSSource];
+                    break;
+                case STPSourceStatusFailed:
+                    if (threeDSecureRequired) {
+                        // If required, fail with error
+                        [self didFinishWithStatus:STPPaymentStatusError
+                                            error:nil]; // TODO: Create STPPaymentContext error domain
+                    }
+                    else {
+                        // If not required, charge original source
+                        self.state = STPPaymentContextStateNone;
+                        [self requestSynchronousSourcePayment:cardSource];
+                    }
+
+                    break;
+                case STPSourceStatusConsumed:
+                case STPSourceStatusChargeable:
+                    // Success, we can just finish
+                    // (charge will happen on webhook)
+                    [self didFinishWithStatus:STPPaymentStatusSuccess
+                                        error:nil];
+                    break;
+                case STPSourceStatusUnknown:
+                case STPSourceStatusCanceled:
+                    [self didFinishWithStatus:STPPaymentStatusError
+                                        error:nil]; // TODO: Create STPPaymentContext error domain
```

Comment:
I think creating a new error domain and all the associated codes is kind of a big diff. Was putting all these todos in so we can easily find where to go back and fill them in, but I'd like to do that in a separate PR instead of grouping it with this one.